### PR TITLE
uplink/core: add wireguard-meshing

### PR DESCRIPTION
**Pull request by PolynomialDivision**

### CODE REVIEW

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+setup_wgtunnel() {
+	if [ ! -f "/root/wg.key" ]; then
+		wg genkey | tee wg.key | wg pubkey >wg.pub
+	fi
+
+	# do some kind of load balancing
+	randomized_ips=$(shuf -e "$ips" | tr "\n" " " && echo "")
+	for ip in "$randomized_ips"; do
+		port=$(timeout 5 ip netns exec uplink wg-client-installer register --ip $ip --user wginstaller --password wginstaller --mtu 1500 --bandwidth 100 | grep Port | awk '{print $2}')
```
Comment (a22sc) [inline roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:91]: why do u limit bandwidth?

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		fi
+		i=$((i + 1))
+	done
+
+	if [ $found -eq 0 ]; then
+		uci add babeld interface
+		uci set babeld.@interface[-1].ifname=$1
+		uci commit
+	fi
+	/etc/init.d/babeld reload
```
Comment (a22sc) [inline roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:142]: why reload if no changes occur?

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		uci set olsrd.@Interface[-1].Mode='ether'
+		uci commit
+	fi
+	/etc/init.d/olsrd reload
+}
+
+neighbors_available() {
+	local phy
+
+	wg_interfaces=$(ip link | grep wg | awk '{print $2}' | sed 's/://')
```
Comment (a22sc) [inline roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:168]: short grep search term might lead to wrong results

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (PolynomialDivision) [reply roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:91]: for now it is just a dummy value.

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (PolynomialDivision) [reply roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:146]: The script is only executed if at least something changed. The interface should be deleted. However, I want to also consider corner cases, where still the olsrd interface is already in the config files (although that should not happen).

File: roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (PolynomialDivision) [reply roles/cfg_openwrt/templates/corerouter/init.d/wgconnect.j2:172]: Thanks that is a good idea, to grep for `grep ' wg_[0-9]*``. Hwoeer is that really an issue to do absolute paths? What happens if grep is installed differently?

File: host_vars/b0russia-core/wireguard.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+wg_connect: true
```
Comment (spolack) [inline host_vars/b0russia-core/wireguard.yml:2]: what is wg_connect? why dont you need it at xhain-core?

File: host_vars/ak36-gw/gateway.yml [hunk 5]
```diff
@@ -77,6 +77,7 @@ tunnel_metric: 1024
 # Tunnel metric 1024 as most internet uplinks will hardly reach 40MBit/s
 
 wireguard_base_prefix: 2001:bf7:750:4002::/64
+wireguard_base_v4prefix: 10.31.145.0/24
```
Comment (spolack) [inline host_vars/ak36-gw/gateway.yml:80]: why dont set that on saarbruecker and ohlauer?

File: roles/cfg_openwrt/files/uplink_gateway/init.d/symlink_mesh_routings [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        ln -s /var/run/babeld /etc/config/babeld
+    fi
+
+    # symlink olsrd
+    if [ ! -L "/etc/config/olsrd" ] && [ -f "/rom/etc/config/olsrd" ]; then
+        cp /rom/etc/config/olsrd /var/run
+        rm /etc/config/olsrd
+        ln -s /var/run/olsrd /etc/config/olsrd
+    fi
+}
```
Comment (spolack) [inline roles/cfg_openwrt/files/uplink_gateway/init.d/symlink_mesh_routings:20]: what happens if i powercycle the gateway inbetween?

File: roles/cfg_openwrt/tasks/fix_permissions.yml [hunk 7]
```diff
@@ -16,3 +16,16 @@
...
+- name: Set permission for symlink_mesh_routings script
+  file:
+    path: "{{ config_path }}/etc/init.d/symlink_mesh_routings"
+    mode: "755"
+  ignore_errors: true
+  when: role == "uplink_gateway"
+
+- name: Set permission for wgconnect script
+  file:
+    path: "{{ config_path }}/etc/init.d/wgconnect"
```
Comment (spolack) [inline roles/cfg_openwrt/tasks/fix_permissions.yml:28]: please put wgconnect in a package in our falter-packages feed

File: roles/cfg_openwrt/templates/common/rc.local.j2 [hunk 8]
```diff
@@ -21,3 +21,12 @@ sed -i '/^ \/etc/d' /usr/lib/opkg/status
 {% for override in sysfs_overrides | default([]) %}
 echo {{ override['value'] }} > {{ override['path'] }}
 {% endfor%}
+
+{% if role == 'uplink_gateway' %}
+# install cleanup script
+test -f /etc/crontabs/root || touch /etc/crontabs/root
+grep -q '/usr/share/wginstaller/wg.sh cleanup_wginterfaces' /etc/crontabs/root || {
+    echo "*/10 * * * *	/usr/share/wginstaller/wg.sh cleanup_wginterfaces" >> /etc/crontabs/root
```
Comment (spolack) [inline roles/cfg_openwrt/templates/common/rc.local.j2:29]: please install that from the package.

File: roles/cfg_openwrt/templates/uplink_gateway/config/wgserver.j2 [hunk 9]
```diff
@@ -8,3 +8,4 @@ config server
     {% endif %}
     option wg_key '{{ wireguard_wg_key }}'
     option wg_pub '{{ wireguard_wg_pub }}'
+    option wg_tmp_key '{{ wireguard_wg_tmp_key }}'
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/config/wgserver.j2:11]: please put some comments in here, describing what those options do.

File: group_vars/all/wireguard.yml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+---
+
+# wg-connect
+wg_connct: false
```
Comment (spolack) [inline group_vars/all/wireguard.yml:4]: what is that?

File: group_vars/all/wireguard.yml [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+wg_wan_ip: ""
+wg_wan_mask: ""
+wg_wan_gw: ""
```
Comment (spolack) [inline group_vars/all/wireguard.yml:7]: missing last line

File: roles/cfg_openwrt/files/uplink_gateway/init.d/symlink_mesh_routings (hunk 6 above)
Comment (PolynomialDivision) [reply roles/cfg_openwrt/files/uplink_gateway/init.d/symlink_mesh_routings:20]: good point.

File: host_vars/ak36-gw/gateway.yml (hunk 5 above)
Comment (PolynomialDivision) [reply host_vars/ak36-gw/gateway.yml:80]: Probably I made a mistake doing a rebase.

Comment (PolynomialDivision) [reply host_vars/ak36-gw/gateway.yml:80]: I did this.

**Comment by spolack** [comment]
i think this is obsolete now. Please reopen if not

**Pull request closed without merging**